lib/wibox/widget/piechart.lua

```diff
@@ -0,0 +1,212 @@
+---------------------------------------------------------------------------
+-- Display percentage in a circle.
+--
+-- Note that this widget makes no attempts to prevent overlapping labels or
+-- labels drawn outside of the widget boundaries.
+--
+--@DOC_wibox_widget_defaults_piechart_EXAMPLE@
+-- @author Emmanuel Lepage Valle
+-- @copyright 2012 Emmanuel Lepage Vallee
+-- @release @AWESOME_VERSION@
+-- @classmod wibox.widget.piechart
+---------------------------------------------------------------------------
+
+local color     = require( "gears.color"       )
+local base      = require( "wibox.widget.base" )
+local beautiful = require( "beautiful"         )
+local util      = require( "awful.util"        )
+local pie       = require( "gears.shape"       ).pie
+local unpack    = unpack or table.unpack -- luacheck: globals unpack (compatibility with Lua 5.1)
+
+local module = {}
+
+local piechart = {}
+
+local function draw_label(cr,angle,radius,center_x,center_y,text)
+    cr:move_to(center_x+(radius/2)*math.cos(angle),center_y+(radius/2)*math.sin(angle))
+    cr:line_to(center_x+(1.5*radius)*math.cos(angle),center_y+(1.5*radius)*math.sin(angle))
+    local x,y = cr:get_current_point()
+    cr:line_to(x+(x>center_x and radius/2 or -radius/2),y)
+    local extents
+
+    -- Support older LGI versions
+    if type(cr.font_extents) == "function" then
+        extents = cr:font_extents()
+    else
+        extents = cr.font_extents
+    end
+
+    cr:move_to(x+(x>center_x and radius/2 + 5 or (-radius/2 - cr:text_extents(text).width - 5)),y+(extents.height/4))
```

Comment:
I'd like to have more explanation for this. Why 5? (likely just some distance between the text and the line, perhaps replacing this with a space character in the text?) Why `extents.height/4`? This feels like some magic numbers that hide some incorrectness.